File: frontend/mobile/ios/NextPhoton/Domain/Entities/Session.swift
// Session.swift
// NextPhoton EduCare - iOS Application
//
// Domain entity representing a tutoring session in NextPhoton.
// Sessions are the core service offering - scheduled meetings between
// educators and learners.

import Foundation

/// Represents a tutoring session in NextPhoton
///
/// A session is a scheduled educational meeting between an educator and
/// one or more learners. Sessions can be one-on-one or group-based.
struct Session: Identifiable, Equatable, Hashable, Sendable {
    // MARK: - Properties

    /// Unique identifier for the session
    let id: String

    /// Title or topic of the session
    let title: String

    /// Detailed description of what will be covered
    var description: String?

    /// Subject area (e.g., "Mathematics", "Physics")
    let subject: String

    /// Specific topic within the subject
    var topic: String?

    /// Educator conducting the session
    let educator: SessionParticipant

    /// List of learners attending the session
    var learners: [SessionParticipant]

    /// Scheduled start time
    let startTime: Date

    /// Scheduled end time
    let endTime: Date

    /// Actual start time (when session was started)
    var actualStartTime: Date?

    /// Actual end time (when session was ended)
    var actualEndTime: Date?

    /// Current status of the session
    var status: SessionStatus

    /// Type of session (online, in-person, hybrid)
    let sessionType: SessionType

    /// Meeting link for online sessions
    var meetingLink: URL?

    /// Physical location for in-person sessions
    var location: SessionLocation?

    /// Maximum number of participants allowed
    let maxParticipants: Int

    /// Price for the session
    let price: Decimal

    /// Currency for the price
    let currency: String

    /// Whether the session is recurring
    let isRecurring: Bool

    /// Recurrence pattern if recurring
    var recurrencePattern: RecurrencePattern?

    /// Notes from the educator
    var educatorNotes: String?

    /// Feedback from participants
    var feedback: [SessionFeedback]?

    /// Session creation timestamp
    let createdAt: Date

    /// Last update timestamp
    var updatedAt: Date

    // MARK: - Computed Properties

    /// Duration of the scheduled session in minutes
    var scheduledDurationMinutes: Int {
        Int(endTime.timeIntervalSince(startTime) / 60)
    }

    /// Actual duration if session has ended
    var actualDurationMinutes: Int? {
        guard let start = actualStartTime, let end = actualEndTime else {
            return nil
        }
        return Int(end.timeIntervalSince(start) / 60)
    }

    /// Whether the session is in the past
    var isPast: Bool {
        endTime < Date()
    }

    /// Whether the session is currently ongoing
    var isOngoing: Bool {
        let now = Date()
        return startTime <= now && endTime >= now && status == .inProgress
    }

    /// Whether the session can be joined
    var canJoin: Bool {
        let now = Date()
        let joinWindowStart = startTime.addingTimeInterval(-15 * 60) // 15 minutes before
        return now >= joinWindowStart && now <= endTime && (status == .scheduled || status == .inProgress)
    }

    /// Whether the session can be cancelled
    var canCancel: Bool {
        status == .scheduled && startTime > Date().addingTimeInterval(24 * 60 * 60) // 24 hours before
    }

    /// Formatted price string
    var formattedPrice: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = currency
        return formatter.string(from: price as NSDecimalNumber) ?? "\(currency) \(price)"
    }

    /// Number of current participants
    var participantCount: Int {
        learners.count + 1 // +1 for educator
    }

    /// Available spots
    var availableSpots: Int {
        max(0, maxParticipants - participantCount)
    }
}

// MARK: - Session Status

/// Status of a tutoring session
enum SessionStatus: String, Codable, CaseIterable, Sendable {
    case scheduled = "SCHEDULED"
    case inProgress = "IN_PROGRESS"
    case completed = "COMPLETED"
    case cancelled = "CANCELLED"
    case noShow = "NO_SHOW"
    case rescheduled = "RESCHEDULED"

    /// Human-readable display name
    var displayName: String {
        switch self {
        case .scheduled: return "Scheduled"
        case .inProgress: return "In Progress"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        case .noShow: return "No Show"
        case .rescheduled: return "Rescheduled"
        }
    }

    /// Color name for UI representation
    var colorName: String {
        switch self {
        case .scheduled: return "blue"
        case .inProgress: return "green"
        case .completed: return "gray"
        case .cancelled: return "red"
        case .noShow: return "orange"
        case .rescheduled: return "yellow"
        }
    }

    /// Icon name for UI
    var iconName: String {
        switch self {
        case .scheduled: return "calendar.badge.clock"
        case .inProgress: return "video.fill"
        case .completed: return "checkmark.circle.fill"
        case .cancelled: return "xmark.circle.fill"
        case .noShow: return "person.crop.circle.badge.xmark"
        case .rescheduled: return "calendar.badge.exclamationmark"
        }
    }
}

// MARK: - Session Type

/// Type of tutoring session
enum SessionType: String, Codable, CaseIterable, Sendable {
    case online = "ONLINE"
    case inPerson = "IN_PERSON"
    case hybrid = "HYBRID"

    /// Human-readable display name
    var displayName: String {
        switch self {
        case .online: return "Online"
        case .inPerson: return "In Person"
        case .hybrid: return "Hybrid"
        }
    }

    /// Icon name for UI
    var iconName: String {
        switch self {
        case .online: return "video.fill"
        case .inPerson: return "person.2.fill"
        case .hybrid: return "person.and.arrowtriangle.right.and.arrowtriangle.left"
        }
    }
}

// MARK: - Session Participant

/// Represents a participant in a session
struct SessionParticipant: Identifiable, Equatable, Hashable, Codable, Sendable {
    /// User ID of the participant
    let id: String

    /// Display name of the participant
    let displayName: String

    /// Avatar URL
    var avatarURL: URL?

    /// Role in the session (educator or learner)
    let role: ParticipantRole

    /// Whether the participant has joined
    var hasJoined: Bool

    /// Join timestamp
    var joinedAt: Date?

    /// Leave timestamp
    var leftAt: Date?

    /// Attendance status
    var attendanceStatus: AttendanceStatus

    enum ParticipantRole: String, Codable, Sendable {
        case educator = "EDUCATOR"
        case learner = "LEARNER"
    }

    enum AttendanceStatus: String, Codable, Sendable {
        case pending = "PENDING"
        case present = "PRESENT"
        case absent = "ABSENT"
        case late = "LATE"
    }
}

// MARK: - Session Location

/// Physical location for in-person sessions
struct SessionLocation: Equatable, Hashable, Codable, Sendable {
    /// Address line 1
    let address: String

    /// City
    let city: String

    /// State/Province
    var state: String?

    /// Country
    let country: String

    /// Postal/ZIP code
    var postalCode: String?

    /// Latitude coordinate
    var latitude: Double?

    /// Longitude coordinate
    var longitude: Double?

    /// Additional location notes
    var notes: String?

    /// Full formatted address
    var formattedAddress: String {
        var parts = [address, city]
        if let state = state {
            parts.append(state)
        }
        parts.append(country)
        if let postalCode = postalCode {
            parts.append(postalCode)
        }
        return parts.joined(separator: ", ")
    }
}

// MARK: - Recurrence Pattern

/// Pattern for recurring sessions
struct RecurrencePattern: Equatable, Hashable, Codable, Sendable {
    /// Frequency of recurrence
    let frequency: RecurrenceFrequency

    /// Interval (e.g., every 2 weeks)
    let interval: Int

    /// Days of week for weekly recurrence
    var daysOfWeek: [DayOfWeek]?

    /// Day of month for monthly recurrence
    var dayOfMonth: Int?

    /// End date for recurrence
    var endDate: Date?

    /// Number of occurrences
    var occurrences: Int?

    enum RecurrenceFrequency: String, Codable, Sendable {
        case daily = "DAILY"
        case weekly = "WEEKLY"
        case biweekly = "BIWEEKLY"
        case monthly = "MONTHLY"
    }

    enum DayOfWeek: Int, Codable, CaseIterable, Sendable {
        case sunday = 1
        case monday = 2
        case tuesday = 3
        case wednesday = 4
        case thursday = 5
        case friday = 6
        case saturday = 7

        var shortName: String {
            switch self {
            case .sunday: return "Sun"
            case .monday: return "Mon"
            case .tuesday: return "Tue"
            case .wednesday: return "Wed"
            case .thursday: return "Thu"
            case .friday: return "Fri"
            case .saturday: return "Sat"
            }
        }
    }
}

// MARK: - Session Feedback

/// Feedback for a completed session
struct SessionFeedback: Identifiable, Equatable, Hashable, Codable, Sendable {
    /// Unique identifier
    let id: String

    /// User who provided feedback
    let userId: String

    /// User's display name
    let userDisplayName: String

    /// Rating (1-5)
    let rating: Int

    /// Written feedback
    var comment: String?

    /// Timestamp
    let createdAt: Date

    /// Whether this is from the educator or learner
    let fromRole: SessionParticipant.ParticipantRole
}

// MARK: - Session Filter

/// Filters for querying sessions
struct SessionFilter: Sendable {
    var status: [SessionStatus]?
    var sessionType: [SessionType]?
    var subject: String?
    var educatorId: String?
    var learnerId: String?
    var startDateFrom: Date?
    var startDateTo: Date?
    var minPrice: Decimal?
    var maxPrice: Decimal?
}

// MARK: - Session Booking Request

/// Request to book a new session
struct SessionBookingRequest: Sendable {
    let educatorId: String
    let subject: String
    var topic: String?
    let startTime: Date
    let endTime: Date
    let sessionType: SessionType
    var meetingLink: URL?
    var location: SessionLocation?
    var notes: String?
}

File: frontend/mobile/android/app/src/main/java/com/nextphoton/domain/entities/Session.kt
/**
 * NextPhoton EduCare - Domain Layer
 * Session Entity - Core domain model for tutoring sessions
 *
 * This entity represents a tutoring session in the NextPhoton platform.
 * Sessions are the core product of NextPhoton, connecting educators
 * with learners for personalized education.
 *
 * Session lifecycle:
 * SCHEDULED -> IN_PROGRESS -> COMPLETED
 *           -> CANCELLED (from any state)
 */
package com.nextphoton.domain.entities

import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId

/**
 * Represents a tutoring session in the NextPhoton platform
 *
 * @property id Unique identifier for the session
 * @property title Session title or subject
 * @property description Detailed description of session content
 * @property subject Academic subject for the session
 * @property educatorId ID of the educator conducting the session
 * @property educatorName Display name of the educator
 * @property educatorAvatarUrl Educator's profile picture URL
 * @property learnerId ID of the learner attending the session
 * @property learnerName Display name of the learner
 * @property scheduledAt Scheduled start time of the session
 * @property duration Duration of the session
 * @property status Current status of the session
 * @property sessionType Type of session (online/in-person)
 * @property meetingUrl URL for online sessions (Zoom, Meet, etc.)
 * @property location Physical location for in-person sessions
 * @property notes Additional notes for the session
 * @property rating Learner's rating of the session (1-5)
 * @property feedback Learner's feedback text
 * @property createdAt Session creation timestamp
 * @property updatedAt Last update timestamp
 */
data class Session(
    val id: String,
    val title: String,
    val description: String? = null,
    val subject: Subject,
    val educatorId: String,
    val educatorName: String,
    val educatorAvatarUrl: String? = null,
    val learnerId: String,
    val learnerName: String,
    val scheduledAt: Instant,
    val duration: Duration = Duration.ofHours(1),
    val status: SessionStatus = SessionStatus.SCHEDULED,
    val sessionType: SessionType = SessionType.ONLINE,
    val meetingUrl: String? = null,
    val location: String? = null,
    val notes: String? = null,
    val rating: Int? = null,
    val feedback: String? = null,
    val createdAt: Instant = Instant.now(),
    val updatedAt: Instant = Instant.now()
) {
    /**
     * Calculated end time of the session
     */
    val endTime: Instant
        get() = scheduledAt.plus(duration)

    /**
     * Checks if the session is happening today
     */
    fun isToday(zoneId: ZoneId = ZoneId.systemDefault()): Boolean {
        val sessionDate = LocalDate.ofInstant(scheduledAt, zoneId)
        val today = LocalDate.now(zoneId)
        return sessionDate == today
    }

    /**
     * Checks if the session is in the future
     */
    val isUpcoming: Boolean
        get() = scheduledAt.isAfter(Instant.now()) && status == SessionStatus.SCHEDULED

    /**
     * Checks if the session is currently in progress
     */
    val isInProgress: Boolean
        get() {
            val now = Instant.now()
            return now.isAfter(scheduledAt) && now.isBefore(endTime) &&
                    status == SessionStatus.IN_PROGRESS
        }

    /**
     * Checks if the session can be cancelled
     */
    val canCancel: Boolean
        get() = status == SessionStatus.SCHEDULED &&
                scheduledAt.isAfter(Instant.now().plus(Duration.ofHours(24)))

    /**
     * Checks if the session can be rescheduled
     */
    val canReschedule: Boolean
        get() = status == SessionStatus.SCHEDULED &&
                scheduledAt.isAfter(Instant.now().plus(Duration.ofHours(48)))

    /**
     * Gets the session time as LocalTime for display
     */
    fun getLocalTime(zoneId: ZoneId = ZoneId.systemDefault()): LocalTime {
        return LocalTime.ofInstant(scheduledAt, zoneId)
    }

    /**
     * Gets the session date as LocalDate for display
     */
    fun getLocalDate(zoneId: ZoneId = ZoneId.systemDefault()): LocalDate {
        return LocalDate.ofInstant(scheduledAt, zoneId)
    }

    /**
     * Formats the duration as a human-readable string
     */
    val formattedDuration: String
        get() {
            val hours = duration.toHours()
            val minutes = duration.toMinutes() % 60
            return when {
                hours > 0 && minutes > 0 -> "${hours}h ${minutes}m"
                hours > 0 -> "${hours}h"
                else -> "${minutes}m"
            }
        }
}

/**
 * Enumeration of session statuses
 */
enum class SessionStatus(val displayName: String) {
    SCHEDULED("Scheduled"),
    IN_PROGRESS("In Progress"),
    COMPLETED("Completed"),
    CANCELLED("Cancelled"),
    NO_SHOW("No Show")
}

/**
 * Enumeration of session types
 */
enum class SessionType(val displayName: String) {
    ONLINE("Online"),
    IN_PERSON("In Person"),
    HYBRID("Hybrid")
}

/**
 * Enumeration of academic subjects
 */
enum class Subject(val displayName: String, val icon: String) {
    MATHEMATICS("Mathematics", "calculate"),
    PHYSICS("Physics", "science"),
    CHEMISTRY("Chemistry", "science"),
    BIOLOGY("Biology", "biotech"),
    ENGLISH("English", "menu_book"),
    HISTORY("History", "history_edu"),
    GEOGRAPHY("Geography", "public"),
    COMPUTER_SCIENCE("Computer Science", "computer"),
    ECONOMICS("Economics", "trending_up"),
    BUSINESS("Business Studies", "business"),
    ART("Art", "palette"),
    MUSIC("Music", "music_note"),
    PHYSICAL_EDUCATION("Physical Education", "fitness_center"),
    FOREIGN_LANGUAGE("Foreign Language", "translate"),
    GENERAL("General", "school")
}

/**
 * Data class for session booking request
 */
data class SessionBookingRequest(
    val educatorId: String,
    val learnerId: String,
    val subject: Subject,
    val title: String,
    val description: String? = null,
    val scheduledAt: Instant,
    val duration: Duration,
    val sessionType: SessionType,
    val notes: String? = null
)

/**
 * Data class for session filter criteria
 */
data class SessionFilter(
    val status: SessionStatus? = null,
    val subject: Subject? = null,
    val educatorId: String? = null,
    val learnerId: String? = null,
    val fromDate: LocalDate? = null,
    val toDate: LocalDate? = null,
    val sessionType: SessionType? = null
)
